File: source/includes/fts-tutorial/date-range/complex-tutorial.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Bson.Serialization.Conventions;
using MongoDB.Driver;
using MongoDB.Driver.Search;

public class DateRangeComplexExample
{
    private const string MongoConnectionString = "<connection-string>";

    public static void Main(string[] args)
    {
        // allow automapping of the camelCase database fields to our MovieDocument
        var camelCaseConvention = new ConventionPack { new CamelCaseElementNameConvention() };
        ConventionRegistry.Register("CamelCase", camelCaseConvention, type => true);

        // connect to your Atlas cluster
        var mongoClient = new MongoClient(MongoConnectionString);
        var mflixDatabase = mongoClient.GetDatabase("sample_mflix");
        var moviesCollection = mflixDatabase.GetCollection<MovieDocument>("movies");

        // declare data for compound query
        var startDate = new DateTime(2015, 01, 01, 0, 0, 0, DateTimeKind.Utc);
        var endDate = new DateTime(2015, 12, 31, 0, 0, 0, DateTimeKind.Utc);
        var nearDate = new DateTime(2015, 07, 01, 0, 0, 0, DateTimeKind.Utc);

        // define and run pipeline
        var results = moviesCollection.Aggregate()
            .Search(Builders<MovieDocument>.Search.Compound()
                .Must(Builders<MovieDocument>.Search.Range(movie => movie.Released, SearchRangeBuilder.Gt(startDate).Lt(endDate)))
                .Should(Builders<MovieDocument>.Search.Near(movie => movie.Released, nearDate, 2629800000))
                .MustNot(Builders<MovieDocument>.Search.Text(movie => movie.Genres, "Documentary")),
                indexName: "date-range-tutorial")
            .Project<MovieDocument>(Builders<MovieDocument>.Projection
                .Include(movie => movie.Genres)
                .Include(movie => movie.Released)
                .Include(movie => movie.Title)
                .Exclude(movie => movie.Id)
                .MetaSearchScore(movie => movie.Score))
            .Limit(6)
            .ToList();

        // print results
        foreach (var movie in results)
        {
            Console.WriteLine(movie.ToJson());
        }
    }
}

[BsonIgnoreExtraElements]
public class MovieDocument
{
    [BsonIgnoreIfDefault]
    public ObjectId Id { get; set; }
    public string [] Genres { get; set; }
    public DateTime Released { get; set; }
    public string Title { get; set; }
    public double Score { get; set; }
}

File: source/includes/fts-tutorial/date-range/complex-tutorial.go
package main

import (
	"context"
	"fmt"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

func main() {
	// connect to your Atlas cluster
	client, err := mongo.Connect(options.Client().ApplyURI("<connection-string>"))
	if err != nil {
		panic(err)
	}
	defer client.Disconnect(context.TODO())

	// set namespace
	collection := client.Database("sample_mflix").Collection("movies")

	// define pipeline stages
	searchStage := bson.D{{Key: "$search", Value: bson.M{
		"index": "date-range-tutorial",
		"compound": bson.M{
			"must": bson.M{
				"range": bson.M{
					"path": "released",
					"gt":   time.Date(2015, time.January, 1, 0, 0, 0, 0, time.UTC),
					"lt":   time.Date(2015, time.December, 31, 0, 0, 0, 0, time.UTC),
				}},
			"should": bson.D{
				{Key: "near", Value: bson.M{
					"path":   "released",
					"origin": time.Date(2015, time.July, 1, 0, 0, 0, 0, time.UTC),
					"pivot":  2629800000,
				}}},
			"mustNot": bson.D{
				{Key: "text", Value: bson.M{
					"path": "genres", "query": "Documentary",
				}}},
		}}}}
	projectStage := bson.D{
		{Key: "$project", Value: bson.D{
			{Key: "_id", Value: 0},
			{Key: "title", Value: 1},
			{Key: "released", Value: 1},
			{Key: "genres", Value: 1},
			{Key: "score", Value: bson.D{
				{Key: "$meta", Value: "searchScore"},
			}},
		}},
	}
	limitStage := bson.D{{Key: "$limit", Value: 6}}

	// run pipeline
	cursor, err := collection.Aggregate(context.TODO(), mongo.Pipeline{searchStage, projectStage, limitStage})
	if err != nil {
		panic(err)
	}

	// print results
	var results []bson.D
	if err = cursor.All(context.TODO(), &results); err != nil {
		panic(err)
	}
	for _, result := range results {
		fmt.Println(result)
	}
}
